### PR TITLE
Add kernel module handler for kvm hugepage support

### PR DESCRIPTION
**Pull request by smitterl**
Handle cases where kvm module is loaded without hugepage support

1. Add class to help load and restore Linux kernel module configuration.

unittests:
```bash
python -m unittest /root/code/avocado-vt/virttest/unittests/test_utils_kernel_module.py -v
test_instance_backs_up (virttest.unittests.test_utils_kernel_module.TestKernelModuleLoaded) ... ok
test_instance_persistence_existing (virttest.unittests.test_utils_kernel_module.TestKernelModuleLoaded) ... ok
test_instance_persistence_not_existing (virttest.unittests.test_utils_kernel_module.TestKernelModuleLoaded) ... ok
test_load_module (virttest.unittests.test_utils_kernel_module.TestKernelModuleLoaded) ... ok
test_load_module_persists_itself_by_default (virttest.unittests.test_utils_kernel_module.TestKernelModuleLoaded) ... ok
test_load_module_twice_different_params (virttest.unittests.test_utils_kernel_module.TestKernelModuleLoaded) ... ok
test_load_module_twice_restore (virttest.unittests.test_utils_kernel_module.TestKernelModuleLoaded) ... ok
test_load_module_twice_same_params (virttest.unittests.test_utils_kernel_module.TestKernelModuleLoaded) ... ok
test_restore (virttest.unittests.test_utils_kernel_module.TestKernelModuleLoaded) ... ok
test_restore_only_if_updated_config (virttest.unittests.test_utils_kernel_module.TestKernelModuleLoaded) ... ok
test_restore_removes_persistence_file_per_default (virttest.unittests.test_utils_kernel_module.TestKernelModuleLoaded) ... ok
test__load_config_module_loaded (virttest.unittests.test_utils_kernel_module.TestModuleInfo) ... ok
test__load_config_module_not_loaded (virttest.unittests.test_utils_kernel_module.TestModuleInfo) ... ok
test_instance_stores_not_loaded (virttest.unittests.test_utils_kernel_module.TestModuleNotLoaded) ... ok
test_load_is_forced (virttest.unittests.test_utils_kernel_module.TestModuleNotLoaded) ... ok
test_load_not_forced (virttest.unittests.test_utils_kernel_module.TestModuleNotLoaded) ... ok

----------------------------------------------------------------------
Ran 16 tests in 0.012s

OK
```

2. Hook activating hugepages with config into env_process
 a. Introduce variable `kvm_module_parameters` to define custom module parameters on load
 b. Introduce variable `kvm_module_force_load` to avoid loading module if not previously loaded
 c. Only load module parameters if not loaded before. Restore original config at start of first following
    test case that does not define `kvm_module_parameters`. `KernelModuleHandler` takes care of only
    loading module parameters if they differ from previous setting in order to reduce stress on test system.
    Use persistence file in avocado test directory to pass module handler instance from one test execution
    to the next during the same job. In order to further reduce clutter in log, only create module handler
    if parameters will be updated or need to be restored.
 d. Sample log output with `kvm_module_parameters = 'hpage=1'` and `kvm_module_force_load = yes`: 4 test
    cases were run, where only 2nd and 3rd test case required parameter update:
```bash
2020-02-19 05:42:51,248 utils_kernel_mod L0103 DEBUG| Backed up kvm module state (was_loaded, params)=(True, halt_poll_ns_shrink=0 hpage=0 halt_po
ll_ns=50000 halt_poll_ns_grow_start=10000 halt_poll_ns_grow=2 nested=0 halt_poll_max_steal=10)
2020-02-19 05:42:51,249 utils_kernel_mod L0056 DEBUG| Loading module: rmmod kvm; modprobe kvm hpage=1
2020-02-19 05:47:32,173 utils_kernel_mod L0045 DEBUG| Not reloading module, same parameters requested for module kvm: hpage=1
2020-02-19 05:48:12,169 utils_kernel_mod L0079 DEBUG| Restoring module state: rmmod kvm; modprobe kvm halt_poll_ns_shrink=0 hpage=0 halt_poll_ns=5
0000 halt_poll_ns_grow_start=10000 halt_poll_ns_grow=2 nested=0 halt_poll_max_steal=10
```

3. Make sure kvm module configuration is restored after last test modified it
   during `cleanup_env`. Catch all exceptions to have rest of cleanup run.

### CODE REVIEW

**Comment by smitterl** [comment]
@dzhengfy @ldoktor

File: virttest/env_process.py [hunk 1]
```diff
@@ -1573,6 +1583,10 @@ def postprocess(test, params, env):
         else:
             _post_hugepages_surp = h.ext_hugepages_surp
 
+    global _kvm_module_handler
+    if _kvm_module_handler:
```
Comment (chunfuwen) [inline virttest/env_process.py]: suggest add judgement ( if not kvm_module_params == "DEFAULT":)to align with _kvm_module_handler before

Comment (ldoktor) [reply virttest/env_process.py]: Well to be honest I prefer this solution as it'd work even when people tackle this module manually (for whatever reason). In case it's defined, let's restore on cleanup...

File: virttest/env_process.py [hunk 2]
```diff
@@ -1573,6 +1583,10 @@ def postprocess(test, params, env):
...
+        _kvm_module_handler.restore()
```
Comment (ldoktor) [inline virttest/env_process.py]: But after you restored you should clean the `_kvm_module_handler` as there are following tests that might use different params and the `restore` would be called multiple times.

File: virttest/utils_kernel_module.py [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+
+    @staticmethod
+    def _to_line(as_dict):
+        """
+        Write dictionary in one line
+
+        :param as_dict: dictionary holding values {key1:value1, key2:value2, ...}
+        :return: string holding values 'key1=value1 key2=value2 ...'
+        """
+        s = ""
```
Comment (ldoktor) [inline virttest/utils_kernel_module.py]: I don't think this function is necessary as it can be a one-liner:

```
" ".join("%s=%s" % _ for _ in as_dict.items())
```

File: virttest/utils_kernel_module.py [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+
+"""Handle Linux kernel modules"""
+
+
+class KernelModuleHandler(object):
+    """Class handling Linux kernel modules"""
+
+    def __init__(self, module_name):
+        self._module_name = module_name
+        self._was_loaded = True
```
Comment (ldoktor) [inline virttest/utils_kernel_module.py]: Well it would be clearer to set this to `None` to indicate it will be detected later in code.

Comment (ldoktor) [reply virttest/utils_kernel_module.py]: ... and obviously set it to True/False there.

File: virttest/utils_kernel_module.py [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+    def load_module(self, force=True, params=""):
+        """
+        Load module with given parameters
+
+        :param force: Force load if currently not loaded. Default is True as restore undoes loading.
+        :param params: Parameters to load with, 'key1=param1 key2=param2 ...'
+        """
+
+        if not force and not self._was_loaded:
+            logging.info("Module %s isn't loaded. Set force=True to force loading." % self._module_name)
```
Comment (ldoktor) [inline virttest/utils_kernel_module.py]: Good, but it'd be really neat to check values of `params` and only warn in case they differ. What do you think?

File: virttest/utils_kernel_module.py [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+        else:
+            cmd = ""
+            if self._was_loaded:
```
Comment (ldoktor) [inline virttest/utils_kernel_module.py]: The same here, this basically results to `rmmod && modprobe` cycle before/after each test, which is a stress test on it's own and might actually influence the testing. How about only doing this in case the specified params differ?

File: virttest/utils_kernel_module.py [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+        if as_dict and len(as_dict) > 0:
+            p = " %s=%s"
+            for k in as_dict:
+                s += p % (k, as_dict[k])
+        return s.strip()
+
+    @staticmethod
+    def _load_config(module_name):
+        """
+        Get current kvm module parameters
```
Comment (ldoktor) [inline virttest/utils_kernel_module.py]: Well the util is generic, even the module name is not tied to KVM, how about leaving the `kvm` word away and leave it as a generic module that, besides others, can serve the KVM module?

File: virttest/env_process.py [hunk 8]
```diff
@@ -1041,6 +1044,13 @@ def preprocess(test, params, env):
                     env["cpu_driver"] = cpu_driver[0]
                     params["cpu_driver"] = env.get("cpu_driver")
 
+    kvm_module_params = params.get("kvm_module_parameters", "DEFAULT")
+    if not kvm_module_params == "DEFAULT":
+        global _kvm_module_handler
+        _kvm_module_handler = KernelModuleHandler('kvm')
+        kvm_module_force = params.get("kvm_module_force_load", "no") == "yes"
```
Comment (ldoktor) [inline virttest/env_process.py]: I can see multiple use-cases here:

1. make sure only my params are set
2. require at least given param

For me the (2) is the best candidate as I want to make sure nested is enabled, but I don't want to reload the module if it already supports this.

The (1) is simpler for cases where I know exactly my params, but it can lead to unexpected module reloads.

Comment (ldoktor) [reply virttest/env_process.py]: Also note that in case of (2) we might argue whether it should only override the `params` and keep the other params as they were, or whether it should use defaults.

I don't really care whether to use defaults or the current values for the undefined params, but I'd really prefer to only reload the module when the `params` differ.

Comment (smitterl) [reply virttest/env_process.py]: > I can see multiple use-cases here:
> 
> 1. make sure only my params are set
> 2. require at least given param
> 
> For me the (2) is the best candidate as I want to make sure nested is enabled, but I don't want to reload the module if it already supports this.
> 
> The (1) is simpler for cases where I know exactly my params, but it can lead to unexpected module reloads.

Though (1) looks like a valuable addition I'd like to defer implementation as out-of-scope for this PR.

File: virttest/env_process.py (hunk 1 above)
Comment (smitterl) [reply virttest/env_process.py]: With latest change this code has been removed.

File: virttest/env_process.py (hunk 2 above)
Comment (smitterl) [reply virttest/env_process.py]: Fixed with last change.

**Comment by smitterl** [comment]
@chunfuwen @ldoktor Thank you for your reviews.
I updated the code in order to have module helper only working if needed, that is, if it has to restore old configuration or if it really has to load a new one.

In addition to unit tests, I tested manually on system level by executing four test cases in order and checked that requirements were met, i.e.

- tc1: did not require module params update
- tc2: did require module params update
- tc3: did require module params update
- tc4: did not require module params update

The tests were executed in the indicated order tc1,tc2,tc3,tc4.
The `kernel_module_handler` log:
```bash
# cat /root/avocado/job-results/job-2020-02-18T12.03-9a8c511/job.log|grep kernel_mod
2020-02-18 12:05:19,986 utils_kernel_mod L0098 DEBUG| Backed up kvm module state (was_loaded, params)=(True, halt_poll_ns_shrink=0 hpage=0 halt_poll_ns=50000 halt_poll_ns_grow_start=10000 halt_po
ll_ns_grow=2 nested=0 halt_poll_max_steal=10)
2020-02-18 12:05:19,986 utils_kernel_mod L0054 DEBUG| Loading module: rmmod kvm; modprobe kvm hpage=1
2020-02-18 12:11:30,088 utils_kernel_mod L0044 INFO | Not reloading module, same parameters requested for module kvm: hpage=1
2020-02-18 12:12:57,193 utils_kernel_mod L0075 DEBUG| Restoring module state: rmmod kvm; modprobe kvm halt_poll_ns_shrink=0 hpage=0 halt_poll_ns=50000 halt_poll_ns_grow_start=10000 halt_poll_ns_g
row=2 nested=0 halt_poll_max_steal=10
```

The case where parameters are different is covered by unit test `test_load_module_twice_different_params`.

File: virttest/utils_kernel_module.py (hunk 6 above)
Comment (ldoktor) [reply virttest/utils_kernel_module.py]: Hello @smitterl, what I meant here is that it'd be nice to check the required params from `params` and in case they don't differ from the actual params from `/sys` just leave it as is (without any backup, store or module reload).

Now let's say you want hugepages support turned on (hpage). The current implementation reloads the module no matter what the original setting was the first time it loads and then keeps it enabled until the end of testing where it reverts to the original setting.

The version suggested by me would only get the `/sys` values, checked the `hpage` (that was specified in `params`) is enabled and proceed with no reloads, no state, no cleanup afterwards. On the other hand if the `hpage` was off, it'd have reload the module for each test that asks for it.

So to wrap it up, my solution should be way simpler, should result in faster execution when no changes are required but would result in reload for each test that requires different params (that especially in CI should be eliminated by custom setup). Your solution currently adds one reload that can be eliminated but more importantly adds the state and potentially corner-cases that might be problematic to spot and debug.

PS: I still haven't reviewed the whole code so maybe it's cleverer than I thought, but so far I don't see any checks for the expected and actual setting...

Comment (ldoktor) [reply virttest/utils_kernel_module.py]: PPS: By the potential corner-cases I mean multiple tests requiring different params etc. I haven't checked the code yet but I guess it might become problematic, which is why I'd prefer stateless implementation... (not sure what params you have in mind, I'm only using nested and hpage and don't combine their execution so setting it in advance)

Comment (smitterl) [reply virttest/utils_kernel_module.py]: @ldoktor As pointed out by you, I think it's a good idea to 
1. check if requested parameters are already set (by comparing with `'cat /sys/module/X/paramters/*'`) - and skip if so
2. add additionally a log to reflect the full module parameters list if the module has been loaded - that is, if not skipping

By adding 1. to the current stateful implementation, we avoid the additional reload you mention. By adding 2. we make sure that full debugging is possible.

The original use case motivating this change is simply to activate hugepage support on s390x which isn't available in our test environments' kernel by default. I have no use cases where several test cases would require different module parameters.

I don't see further advantage of having a stateless implementation as originally proposed: if huge page support is not available by default, as in the original use case, and if the support should be active in each test in a sequence, the module load and restore commands are called only once per sequence which still reduces stress on the system and test execution times.

Therefore, my proposal would be to add 1. and 2. to the current implementation. Please, let me know.

File: virttest/utils_kernel_module.py [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+                                                    self._module_name,
+                                                    self._config_backup)
+            logging.debug("Restoring module state: %s" % cmd)
+            status, output = process.getstatusoutput(cmd, shell=True,
+                                                     ignore_status=True)
+            if status:
+                raise exceptions.TestError("Couldn't restore module %s: %s" % (
+                    self._module_name, self._config_backup))
+
+        if self._persistence_file and \
```
Comment (ldoktor) [inline virttest/utils_kernel_module.py]: in avocado we use brackets and not escaped \ns (it's somewhere in code-style guide...)

File: virttest/utils_kernel_module.py [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+        else:
+            self._was_loaded = False
+        logging.debug("Backed up %s module state (was_loaded, params)"
+                      "=(%s, %s)" % (self._module_name, self._was_loaded,
+                                     self._config_backup))
+
+    def _load(self):
+        try:
+            with open(self._persistence_file, 'rb') as pickled:
+                self.__dict__.update(pickle.load(pickled))
```
Comment (ldoktor) [inline virttest/utils_kernel_module.py]: Suggestion: How about to dump only the `_config_backup` and `_was_loaded` as we don't really care about the others? Would be more explicit...

File: virttest/utils_kernel_module.py [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+        Create kernel module handler. It backs up the current status
+        of the kernel module or get its current status from a backed up
+        file.
+
+        :param module_name:
+        :param persistence_file: absolute file path to store object in
+        """
+        self._module_name = module_name
+        self._was_loaded = None
+        self._loaded_config = ""
```
Comment (ldoktor) [inline virttest/utils_kernel_module.py]: I know in C mixing strings and Nonetype is not recommended (and hard), but in python it's the usual way to distinguish between not defined and empty string. I don't say it's a mistake, just a different from the usual...

File: avocado_vt/test.py [hunk 12]
```diff
@@ -70,10 +71,25 @@ def cleanup_env(env_filename, env_version):
...
     env.destroy()
 
 
+def try_restore_kvm_module_parameters():
+    """Restores original kvm module parameters if modified by last test"""
+
+    try:
+        kvm_module_handler_file = env_process.KVM_MODULE_HANDLER_FILE
+        if os.path.isfile(kvm_module_handler_file):
+            utils_kernel_module.\
```
Comment (ldoktor) [inline avocado_vt/test.py]: again, we use brackets instead of escaped newlines...

**Comment by smitterl** [comment]
> I gave it several thoughts and I still prefer stateless implementation, but this one does seem to cover all potential issues I had in mind and looks pretty solid so feel free to keep the state. The only remaining issue I see (apart from little nitpicks here and there) is the reload when the requested params are already matching.
> 
> Oh and I forgot to add, please update the commit message not to include the selftests. Commit message should express what are you changing/implementing, why and eventually some useful high-level details.

@ldoktor Thank you very much for reviewing it. I'm not sure which is the best solution. But I believe you have a valid point by pointing out a) less complexity of implementation and b) that CI deployment should better take care of long running module setups.

Therefore, I also created https://github.com/avocado-framework/avocado-vt/pull/2438 with all your suggestions. Should we close this PR in favor of it? Please, let me know. Thanks!

**Comment by smitterl** [comment]
Closing this in favor of https://github.com/avocado-framework/avocado-vt/pull/2438
